doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -3636,6 +3636,236 @@ rewriting rule is stated using Leibniz equality (as opposed to setoid
 relations). It will be extended to other rewriting relations in the
 future.
 
+.. _under_over_ssr:
+
+A convenience under tactic
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Goals involving objects defined with higher-order functions often
+require "rewriting under binders". While setoid rewriting is a
+possible approach in this case, it is common to use regular rewriting
+along with dedicated extensionality lemmas. This may cause some
+practical issues during the development of the corresponding scripts,
+notably as we might be forced to provide the rewrite tactic with
+complete terms, as shown by the simple example below.
+
+.. example::
+
+   .. coqtop:: reset none
+
+      From Coq Require Import ssreflect.
+      Set Implicit Arguments.
+      Unset Strict Implicit.
+      Unset Printing Implicit Defensive.
+
+   .. coqtop:: in
+
+      Axiom subnn : forall n : nat, n - n = 0.
+      Parameter G : (nat -> nat) -> nat -> nat.
+      Axiom eq_G :
+        forall F1 F2 : nat -> nat,
+        (forall n : nat, F1 n = F2 n) ->
+        forall n : nat, G F1 n = G F2 n.
+
+   .. coqtop:: all
+
+      Lemma example_G (n : nat) : G (fun n => n - n) n >= 0.
+
+   In this context, one cannot directly use ``eq_G``:
+
+   .. coqtop:: all fail
+
+      rewrite eq_G.
+
+   as we need to explicitly provide the non-inferable argument ``F2``,
+   which corresponds here to the term we want to obtain *after* the
+   rewriting step. We can thus do:
+
+   .. coqtop:: all abort
+
+      rewrite (@eq_G _ (fun _ : nat => 0)); last first.
+        by move=> m; rewrite subnn.
+
+   Otherwise, a more convenient approach consists in relying on the
+   ``under`` tactic that alleviates this bookkeeping:
+
+   .. coqtop:: all abort
+
+      Lemma example_G (n : nat) : G (fun n => n - n) n >= 0.
+      under m: eq_G by rewrite subnn.
+
+  ``(* FIXME: replace with do... *)``
+
+The ``under`` tactics enjoys two different modes (the "interactive
+mode", and the "one-liner mode" that was mentioned in the example
+above). The corresponding syntax is as follows:
+
+.. tacn:: under {? @occ_switch} {? [ @r_pattern ] } @term {? => @i_item}
+   :name: under
+
+   rewrite under ``term`` in :ref:`interactive mode <under_interactive>`
+
+.. tacn:: under {? @occ_switch} {? [ @r_pattern ] } @term {? => @i_item} do ( @tactic | [ {*| @tactic } ] )
+   :name: under-do
```

Comment:
I don't think it needs a specific index entry given how close the two are. If it needs one, the practice would rather be `under ... do`. But can't you just merge the two grammars?

```
.. tacn:: under {? @occ_switch} {? [ @r_pattern ] } @term {? => @i_item} {? do ( @tactic | [ {*| @tactic } ] ) }

   Rewrite under :token:`term`. We distinguish two modes,
   :ref:`interactive mode <under_interactive>` without a `do` clause, and
   :ref:`one-liner mode <under_one_liner>` with a `do` clause, which are
   explained in more details below.
```